pycqed/analysis_v3/data_processing.py

```diff
@@ -577,318 +440,136 @@ def rotate_iq(data_dict, data_keys_in, data_keys_out, **params):
     return data_dict
 
 
-def rotate_1d_array(data_dict, data_keys_in, data_keys_out, **params):
+def rotate_1d_array(data_dict, keys_out, keys_in=None, **params):
     """
     Rotates 1d array based on information in the CalibrationPoints objects.
     The number of CalibrationPoints objects should equal the number of
-    key names in data_keys_in.
+    key names in keys_in.
     :param data_dict: OrderedDict containing data to be processed and where
                 processed data is to be stored
-    :param data_keys_in: list of key names or dictionary keys paths in
+    :param keys_in: list of key names or dictionary keys paths in
                 data_dict for the data to be processed
-    :param data_keys_out: list of key names or dictionary keys paths in
+    :param keys_out: list of key names or dictionary keys paths in
                 data_dict for the processed data to be saved into
     :param params: keyword arguments.:
         cal_points_list (list): list of CalibrationPoints objects.
         last_ge_pulses (list): list of booleans
 
     Assumptions:
         - one output key per input key
-        - len(data_keys_in) == len(data_keys_out)
+        - len(keys_in) == len(keys_out)
         - cal_points_list exists in **params
-        - len(cp_list) == len(data_keys_in)
+        - len(cp_list) == len(keys_in)
         - one CalibrationPoints object input key
         - assumes the dicts returned by CalibrationPoints.get_indices(),
         CalibrationPoints.get_rotations() are keyed by channel number strings;
         ex: indices for ch 0: {'0': {'g': [-4, -3], 'e': [-2, -1]}}
     """
-    cp_list = get_param('cal_points_list', data_dict, **params)
-    if cp_list is not None:
-        cp_list = [eval(cp) for cp in cp_list]
-    else:
-        raise ValueError('cal_points object not found.')
-
-    last_ge_pulses = get_param('last_ge_pulses', data_dict, default_value=[],
-                               **params)
-    meas_obj_value_names_map = get_param('meas_obj_value_names_map', data_dict,
-                                         **params)
-    if meas_obj_value_names_map is None:
-        raise ValueError('Unknown meas_obj_value_names_map.')
-
-    data_to_proc_dict = get_data_to_process(data_dict, data_keys_in)
-    if len(data_keys_out) != len(data_to_proc_dict):
-        raise ValueError('data_keys_out and data_keys_in do not have '
+    data_to_proc_dict = help_func_mod.get_data_to_process(data_dict, keys_in)
+    keys_in = list(data_to_proc_dict)
+    if len(keys_out) != len(data_to_proc_dict):
+        raise ValueError('keys_out and keys_in do not have '
                          'the same length.')
-    if len(cp_list) != len(data_to_proc_dict):
-        raise ValueError('cal_points_list and data_keys_in do not have '
+
+    cp_list = help_func_mod.get_param('cal_points_list', data_dict, **params)
```

Comment:
I'm not sure I understand why you would need a list of calibration point objects. The cal point object handles multi qubits. I don't think we should create more than 1 per experiment, or am I misunderstanding something ?

But if we keep it for now, does this mean that we have to recreate all the calpoint objects in the notebook if we want to do the analysis in the notebook? of course it's fine for now, but is that the end goal we want ?

Reply:
Good point, cp should not be a list. I will address this when I do a pull request from analysis_v3. 

You would not need to provide the cp in the notebook if they exist in the metadata. It will take them from there and convert them to a list.